File: DevSupportAndroid/src/main/java/com/sup/dev/android/views/elements/cards/Card.java
package com.sup.dev.android.views.elements.cards;

import android.content.Context;
import android.support.annotation.LayoutRes;
import android.view.View;

import com.sup.dev.android.tools.ToolsView;
import com.sup.dev.android.views.adapters.recycler_view.RecyclerCardAdapter;

public abstract class Card {

    public RecyclerCardAdapter adapter;

    public Object tag;

    public void update() {
        if (adapter == null) return;

        View view = adapter.getView(this);

        if (view != null) bindView(view);
    }

    //
    //  Bind
    //

    @LayoutRes
    public abstract int getLayout();

    public abstract void bindView(View view);

    //
    //  Adapter
    //

    public View instanceView(Context context) {
        return ToolsView.inflate(context, getLayout());
    }

    public void setAdapter(RecyclerCardAdapter adapter) {
        this.adapter = adapter;
    }


}

File: src/main/java/com/sup/dev/android/views/cards/Card.kt
package com.sup.dev.android.views.cards

import android.content.Context
import android.view.View
import com.sup.dev.android.app.SupAndroid
import com.sup.dev.android.tools.ToolsView
import com.sup.dev.android.views.support.adapters.CardAdapter


abstract class Card(
    private val layout:Int
) {

    var adapter: CardAdapter? = null
    private var view:View? = null
    var isBinding = false

    var tag: Any? = null

    //
    //  Bind
    //

    fun update() {
        if(isBinding) return
        val view = getView()
        if (view != null) bindCardView(view)
    }

    fun bindCardView(view: View) {
        isBinding = true
        this.view = view
        view.tag = this
        bindView(view)
        isBinding = false
    }

    open fun bindView(view: View) {
    }

    protected open fun instanceView(): View {
        return View(SupAndroid.appContext)
    }

    fun getView(): View? {
        val view = if(adapter == null)  null else adapter!!.getView(this)
        if(view != null){
            this.view = view
            view.tag = this
        }

        if(this.view != null && this.view!!.tag != this) this.view = null

        return this.view
    }

    fun remove(){
        if(adapter != null) adapter!!.remove(this)
    }

    //
    //  Adapter
    //

    open fun instanceView(context: Context): View {
        return if (layout > 0) ToolsView.inflate(context, layout) else instanceView()
    }

    open fun setCardAdapter(adapter: CardAdapter?) {
        this.adapter = adapter
    }


}
